### PR TITLE
refactor whole lib to typescript

### PR DESCRIPTION
**Pull request by bhavya3024**
@vividvilla @ranjanrak 
I refactored the lib to typescript and changed some config on testing files.
But I haven't developed any interfaces yet.

### CODE REVIEW

File: lib/connect.ts [hunk 1]
```diff
@@ -106,9 +106,9 @@ var KiteConnect = function(params) {
         "orders": "/orders",
         "trades": "/trades",
         "order.info": "/orders/{order_id}",
-        "order.place": "/orders/{variety}",
-        "order.modify": "/orders/{variety}/{order_id}",
-        "order.cancel": "/orders/{variety}/{order_id}",
+        "order.place": "/orders/{constiety}",
```
Comment (anurag-roy) [inline lib/connect.ts:109]: This also got replaced while replacing `var` with `const`

File: lib/connect.ts [hunk 2]
```diff
@@ -259,27 +259,27 @@ var KiteConnect = function(params) {
     */
     self.ORDER_TYPE_SL = "SL";
 
-    // Varities
+    // constities
```
Comment (anurag-roy) [inline lib/connect.ts:262]: Same here

File: lib/connect.ts (hunk 1 above)
Comment (bhavya3024) [reply lib/connect.ts]: Done!

**Comment by vividvilla** [comment]
Great work @bhavya3024  👍🏽 thanks for the PR and thanks @anurag-roy for the review.  I haven't done any Typescript in a while so give me some time to go through this. cc @ranjanrak

File: lib/connect.ts [hunk 3]
```diff
@@ -1,10 +1,10 @@
...
 
-var axios = require("axios");
-var csvParse = require("papaparse");
-var sha256 = require("crypto-js/sha256");
-var querystring = require("querystring");
-var utils = require("./utils");
+import axios from "axios";
+import csvParse from "papaparse";
+import sha256 from "crypto-js/sha256";
+const querystring = require("querystring");
```
Comment (anurag-roy) [inline lib/connect.ts:6]: Can be converted to an import statement.
Will need to install `@types/node` as a devDependency.

File: lib/connect.ts [hunk 4]
```diff
@@ -260,7 +260,7 @@ var KiteConnect = function(params) {
     */
     self.ORDER_TYPE_SL = "SL";
 
-    // Varities
+    // constities
```
Comment (anurag-roy) [inline lib/connect.ts:263]: *varieties

File: lib/connect.ts [hunk 5]
```diff
@@ -1341,38 +1341,38 @@ regular).
...
 
-    function request(route, method, params, responseType, responseTransformer, isJSON, queryParams) {
+    function request(route, method, params, responseType?, responseTransformer?, isJSON?, queryParams?) {
         // Check access token
         if (!responseType) responseType = "json";
-        var uri = routes[route];
+        let uri = routes[route];
 
-        // Replace variables in "RESTful" URLs with corresponding params
+        // Replace constiables in "RESTful" URLs with corresponding params
```
Comment (anurag-roy) [inline lib/connect.ts:1365]: *variables

File: lib/utils.ts [hunk 6]
```diff
@@ -1,4 +1,4 @@
-var packageInfo = require('../package.json');
+const packageInfo = require('../package.json');
```
Comment (anurag-roy) [inline lib/utils.ts:1]: Can be converted to an import statement.
Will need `resolveJsonModule: true` in tsconfig.json

File: package.json [hunk 7]
```diff
@@ -2,10 +2,10 @@
   "name": "kiteconnect",
   "version": "4.1.0",
   "description": "The official JS client library for the Kite Connect trading APIs",
-  "main": "./lib/",
+  "main": "./dist/",
   "scripts": {
     "generate-docs": "node_modules/.bin/jsdoc --configure .jsdoc.json --verbose",
```
Comment (anurag-roy) [inline package.json:7]: `generate-docs` script does not work anymore. Probably the target folder in `.jsdoc.json` needs to be updated.

Also, there is no complication script, which should be there maybe?

**Comment by bhavya3024** [comment]
@anurag-roy I never meant to add types or interfaces here as I mentioned in description as well,  I want to see whether all unit test cases work here or not. First migration in Typescript without any interfaces should pass all the test cases which somehow I couldn't able to do at local files because of no mock data was present.
@vividvilla can you enable checks for my PR ? These must pass before adding any interfaces or clubbing any interfaces from @anurag-roy!

**Comment by anurag-roy** [comment]
@bhavya3024 Ok, makes sense now!

**Comment by vividvilla** [comment]
I have glanced through the PR and here are some observations 
- should be compatible with Node v8+ so target should be `es2015`
- missing build command in `package.json`

@anurag-roy has also pointed out few things like converting `querystring` to import statement and broken docs generate command. 

Apart from these the PR looks great, I will run this locally and publish a beta release soon. For mocks you can import this repo as submodule - https://github.com/zerodha/kiteconnect-mocks like we use it on https://github.com/zerodha/gokiteconnect, also you can refer Gokiteconnect lib for types (cc @anurag-roy).

**Comment by bhavya3024** [comment]
@vividvilla I was not able to find the alternative to jsdoc for typescript, which works exactly as jsdoc does for now. rest of the changes I have done, including the interfaces made by @anurag-roy.
Also I thought of converting functions into class, but it was pushing myself to change arguments of many functions leading to change in unit tests, so I avoided it in the first place.

**Comment by bhavya3024** [comment]
@vividvilla I have added TypeDoc to generate docs but it generates docs in a seperate folder. See if it works for you at this time.

**Comment by vividvilla** [comment]
Code seems fine and I have created an alpha NPM release `4.2.0-alpha.0` - https://www.npmjs.com/package/kiteconnect/v/4.2.0-alpha.0 and tests are also passing through. 

@ranjanrak can you take a look at this once? if all good will merge to upstream and create a beta release.

**Review by anurag-roy — APPROVED** [review]
Just installed the alpha version `4.2.0-alpha.0` and took it for a test run. Looks great for the beta release! Will look into improving the typings even more after that.

Type autocompletion works perfectly:

Screenshot 2023-03-13 at 11 36 38 PM

Screenshot 2023-03-13 at 11 37 30 PM

Great work! @bhavya3024

**Comment by ranjanrak** [comment]
Hi @bhavya3024 
I think the doc won't be rendered properly here as you are not following the TSDoc standardized comments notation. You can refer to the [TSDoc official documentation here](https://tsdoc.org/pages/tags/alpha/).
A simple example for reference(here `setAccessToken` function) would be:
```
/**
* Set access_token to the kite connect instance
* @remarks
* This method is part of the KiteConnect package
* @param {string} access_token Token obtained in exchange for `request_token`.
* Once you have obtained `access_token`, you should persist it in a database or session to pass
* to the Kite Connect class initialisation for subsequent requests.
* @public
*/
public setAccessToken(access_token:string):void {
  this.access_token = access_token;  
};
 ```
Once you are done with the comment addition, we will be able to merge this PR. And later PRs can be on types and interfaces addition.

**Comment by ranjanrak** [comment]
@bhavya3024 @anurag-roy 
I feel the end goal for typescript migration should be generating doc using [typedoc](https://typedoc.org/#quick-start), but for this PR we can go ahead with the jsdoc generation for now. Once we have added proper types and interfaces, we can fine-tune all comments as per tsdoc(this can be a later PR).

**Comment by vividvilla** [comment]
@bhavya3024 can you remove the `docs` folder from PR? we don't commit docs in master branch.

**Comment by bhavya3024** [comment]
@vividvilla I have deleted the docs folder

File: lib/connect.ts [hunk 8]
```diff
@@ -170,7 +170,7 @@ var KiteConnect = function(params) {
     requestInstance.interceptors.response.use(function(response) {
         if (self.debug) console.log(response);
 
-        var contentType = response.headers["content-type"];
+        const contentType = response.headers["content-type"];
         if (contentType === "application/json" && typeof response.data === "object") {
```
Comment (gajjardarshithasmukhbhai) [inline lib/connect.ts:174]: @bhavya3024 Just one Suggestion, can we make one constant file and maintain this string constant
ex. `if (contentType === "application/json" && typeof response.data === CONSTANT.APPLICATION_JSON)`

File: lib/connect.ts [hunk 9]
```diff
@@ -466,14 +466,14 @@ var KiteConnect = function(params) {
...
-            var p = _post("api.token", {
+            const checksum = sha256(self.api_key + request_token + api_secret).toString();
+            const p = _post("api.token", {
                     api_key: self.api_key,
                     request_token: request_token,
                     checksum: checksum
                 }, null, formatGenerateSession);
 
-            p.then(function(resp) {
+            p.then(function(resp: any) {
```
Comment (gajjardarshithasmukhbhai) [inline lib/connect.ts:476]: @bhavya3024 Please Avoid `any` instead of Unknown or appropriate Type

File: lib/connect.ts [hunk 10]
```diff
@@ -514,15 +514,15 @@ var KiteConnect = function(params) {
...
 
-            var p = _post("api.token.renew", {
+            const p = _post("api.token.renew", {
                 api_key: self.api_key,
                 refresh_token: refresh_token,
                 checksum: checksum
             });
 
-            p.then(function(resp) {
+            p.then(function(resp: any) {
```
Comment (gajjardarshithasmukhbhai) [inline lib/connect.ts:525]: same here @bhavya3024, let me know if you need any help

File: lib/connect.ts [hunk 11]
```diff
@@ -1098,7 +1098,7 @@ regular).
             trigger_values: params.trigger_values,
             last_price: parseFloat(params.last_price)
         }
-        let orders = []
+        let orders = [] as any[];
```
Comment (gajjardarshithasmukhbhai) [inline lib/connect.ts:1101]: instead of `any[]` pls user like this `let orders: string[] = []` if values are in array.

File: lib/connect.ts [hunk 12]
```diff
@@ -1195,8 +1195,8 @@ regular).
             throw new Error("Invalid postback data or api_secret");
         }
 
-        var inputString = postback_data.order_id + postback_data.order_timestamp + api_secret;
-        var checksum;
+        const inputString = postback_data.order_id + postback_data.order_timestamp + api_secret;
```
Comment (gajjardarshithasmukhbhai) [inline lib/connect.ts:1198]: use string literals over here @bhavya3024

File: lib/connect.ts [hunk 13]
```diff
@@ -1319,8 +1319,8 @@ regular).
     function transformMFInstrumentsResponse(data, headers) {
         // Parse CSV responses
         if (headers["content-type"] === "text/csv") {
-            var parsedData = csvParse.parse(data, {"header": true}).data;
-            for (var item of parsedData) {
+            const parsedData = csvParse.parse(data, {"header": true}).data;
+            for (const item of parsedData as any) {
```
Comment (gajjardarshithasmukhbhai) [inline lib/connect.ts:1323]: If we use @bhavya3024 any all the places no need to use `TypeScript`. if we use appropriate Types So we will track all our objects for particular objects.

File: lib/connect.ts [hunk 14]
```diff
@@ -1341,38 +1341,38 @@ regular).
...
+    function _post(route, params, responseType?, responseTransformer?, isJSON = false, queryParams: any =null) {
         return request(route, "POST", params || {}, responseType, responseTransformer, isJSON, queryParams);
     }
 
-    function _put(route, params, responseType, responseTransformer, isJSON = false, queryParams=null) {
+    function _put(route, params, responseType?, responseTransformer?, isJSON = false, queryParams=null) {
         return request(route, "PUT", params || {}, responseType, responseTransformer, isJSON, queryParams);
     }
 
-    function _delete(route, params, responseType, responseTransformer, isJSON = false) {
```
Comment (gajjardarshithasmukhbhai) [inline lib/connect.ts:1356]: One piece of Advice, we can make a custom hook like `useApi` to handle all these kinds of API methods.

File: lib/connect.ts [hunk 15]
```diff
@@ -1396,7 +1396,7 @@ regular).
 
         // Send auth token
         if (self.access_token) {
-            var authHeader = self.api_key + ":" + self.access_token;
+            const authHeader = self.api_key + ":" + self.access_token;
```
Comment (gajjardarshithasmukhbhai) [inline lib/connect.ts:1399]: string literals please @bhavya3024

**Comment by gajjardarshithasmukhbhai** [comment]
Hi Team, I see your code Just One suggestion we need set at least 2 pr need to be approved before going on approval directly, and then after the code will break in production.

**Comment by vividvilla** [comment]
@gajjardarshithasmukhbhai thanks for the suggestion but that won't be necessary as of now, we can assign approvers automatically on PR but not a fan of letting it automatically merge.

**Comment by anurag-roy** [comment]
Yay! Lets go 🎉

**Pull request merged**